src/utils/diagnostics.ts

```diff
@@ -313,6 +316,66 @@ const diagnoseCardKeys = async (
   return diagnostics;
 };
 
+/**
+ * Provide diagnostics related to variables within an extension.
+ * Users are warned if variables are defined but are not referenced.
+ * Errors are raised if variables are referenced that are not defined.
+ * @param document text document where diagnostics should be applied
+ * @param extension extension.yaml serialized as object
+ * @returns list of diagnostics
+ */
+const diagnoseVariables = async (
+  document: vscode.TextDocument,
+  extension: ExtensionStub,
+): Promise<vscode.Diagnostic[]> => {
+  const content = document.getText();
+  const diagnostics: vscode.Diagnostic[] = [];
+
+  // Honor the user's settings
+  if (
+    !vscode.workspace
+      .getConfiguration("dynatraceExtensions", null)
+      .get("diagnostics.variables", false) as boolean
+  ) {
+    return [];
+  }
+
+  const refVars: Array<string> = [];
+  const defVars: Array<string> = [];
+  const varRegEx = /(var:[a-zA-z.-_]+)/gm;
```

Comment:
I'd still tighten this up a bit. Since colons regularly appear in metric selectors, given the right metric key we could still end up over matching:
![image](https://github.com/dynatrace-extensions/dynatrace-extensions-vscode/assets/16157796/a96bb63c-b482-4e4b-80d0-3b4dd52d559d)

I liked the leading space. Since we know vars are always used as the only value in the yaml attribute, we can bet on `: ` preceding it. So I would use this, with the optional `'` or `"` inbetween it and the var:
`: ['"]?(var:[a-zA-z.-_]+)`

The only scenario where this will fail is if someone very poorly chooses an entity type name to be prefixed with `var`, e.g. `var:some_entity_name` 😄